cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll

```diff
@@ -1216,6 +1218,19 @@ class TranslatedRangeBasedForStmt extends TranslatedStmt, ConditionContext {
     or
     child = this.getUpdate() and
     result = this.getCondition().getFirstInstruction(kind)
+    or
+    exists(int destructorId |
+      destructorId >= this.getFirstDestructorCallIndex() and
+      child = this.getChild(destructorId) and
+      result = this.getChild(destructorId + 1).getFirstInstruction(kind)
+    )
+    or
+    exists(int lastDestructorIndex |
+      lastDestructorIndex =
+        max(int n | exists(this.getChild(n)) and n >= this.getFirstDestructorCallIndex()) and
+      child = this.getChild(lastDestructorIndex) and
+      result = this.getParent().getChildSuccessor(this, kind)
+    )
```

Comment:
I basically did the same thing as for for-loops. What I don't quite understand is why this case is needed. Before this change `result = this.getParent().getChildSuccessor(this, kind)` did not occur in this predicate and still everything worked.

Reply:
This was handled by `result = this.getParent().getChildSuccessor(this, kind)` in `getChildFalseSuccessor` before your change. That is, the last element would be the `this.getCondition()`, and from there we'd have a false successor to `this.getParent().getChildSuccessor(this, kind)`.

Now that you're adding the destructor call after the false edge we need to make sure that we actually step to the parent after those destructor calls (as you're correctly doing here).

Reply:
Hm, yeah now that you mention it. We only call `getLastChild` when it also holds that we don't handle destructors explicitly.

`getLastChild` is invoked recursively through the other overrides of that predicate, though. So I think I still need to convince myself that we can remove it safely in this case, though 🤔

Reply:
In any case, the current implementation matches the one in `ForStmt` (via inheritance of `LoopStmt`), which also has explicit destructor handling.